File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:

        # use frequency map

        # 
        n = len(s)
        l = 0
        r = 0
        freq = {}
        highestFreq = 0
        ans = 0
        while r<n:
            freq[s[r]] = 1 + freq.get(s[r],0)
            highestFreq = max(highestFreq,freq[s[r]])

            if r-l+1 - highestFreq > k:
                while r-l+1 - highestFreq > k:
                    freq[s[l]] -= 1
                    l += 1
                
            ans = max(ans,r-l+1)
            r += 1

        return  ans


        

File: Data Structures & Algorithms/car-fleet/submission-2.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        // sort both arrays by position decreasing to increasing
        if(speed.size() <= 1){
            return 1;
        }
        vector<pair<int,int>> pairs;
        for(int i = 0 ; i<speed.size();i++){
            pairs.push_back({position[i],speed[i]});
        }

        sort(pairs.begin(),pairs.end(), [](const auto& a, const auto& b){
            return a.first > b.first;
        });

        stack<float> fleet;
        float max = 0;
        int ans = 0;
        for(int i = 0 ; i<pairs.size(); i++){
            int distance = target - pairs[i].first;
            float time = float(distance) / pairs[i].second;
            if(time>max){
                max=time;
                ans++;
            }
        }
 
        return ans;
    }
};
